### ISSUE TITLE
Add a way to run some edm::Streams in GPU and the rest in CPU

### ISSUE THREAD

**Issue by makortel** [cms-patatrack/cmssw#64]
Currently the framework runs everything it can on a GPU. When increasing the number of edm::Streams, I believe the first thing to fail is device memory allocation in `beginStreamGPUCuda()`, at which point the job will crash on `cudaCheck()`. We should find a reasonable way to improve for the demonstrator.

1. If our memory allocation model would be more dynamic (=each event allocates necessary memory per-event), and running on module on CPU when GPU memory runs out, we should get the GPU-vs-CPU load balancing automatically.
   * The price is either per-event-`cudaMalloc()`'s or our own device memory allocator
   * When running close to the device memory limit I'd expect the amount of CPU<->GPU transfers to increase (because modules "randomly" fail to get their GPU memory)

2. We can add a configuration parameter to Raw2Cluster (or more specifically to the `GPUCuda` component) specifying the maximum number of `edm::Stream`s to be run on GPU. ~~Configuring it on Raw2Cluster is sufficient, as the rest of the chain will be run on the device where the first product resides.~~
   * Very easy to implement
   * Behaviour is easy to understand and predict
   * doing it only for the raw2cluster will not prevent the other modules to allocate the GPU memory
   * Tuning of the parameter is left to the user
      * how well the approach scales (or doesn't) if more modules become GPU-aware?
      * it does not adjust automatically

3. We could make the `beginStreamGPUCuda()` to detect if device memory allocation fails (easiest to do by making all allocations with `cuda::memory::device::make_unique()` and catching the exception). If the module instance is unable to allocate all its memory, run it on CPU
   * Relatively easy to implement
   * Provides automation
   * I believe we can not make any assumptions on which order the module's `beginStream()` get called. This means that when the device memory runs out, it can happen for edm::Stream` N that e.g.
      * Raw2Cluster and RecHit get their device memory and are thus run on GPU
      * Doublets don't get its memory and is thus forced to run on CPU, leading to transfer of hits to CPU
      * CA did get its device memory and could in principle run on GPU, but the input doublets are on CPU so CA runs on CPU as well. The GPU memory is thus wasted.

@fwyzard @felicepantaleo @VinInn @rovere @vkhristenko

**Comment by cmsbot** [comment]
A new Issue was created by  @makortel Matti Kortelainen.

 can you please review it and eventually sign/assign? Thanks.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-issues.html">here</a>

**Comment by makortel** [comment]
If we take the route of #65 for conditions, running close to the GPU device memory limit becomes a bit dangerous, because (in general) the `acquireGPUCuda()` starts to fail on memory allocations.

I'm leaning towards option 2, which is simple and predictable.

**Comment by VinInn** [comment]
I think option 2 is indeed the simplest to implement and manage at the moment.
It is also an easy, ~reproducible way to test Heterogeneity...

**Comment by fwyzard** [comment]
The problems with 2. are
  - how well the approach scales (or doesn't) if more modules become GPU-aware
  - doing it only for the raw2cluster will not prevent the other modules to allocate the GPU memory
  - it does not adjust automatically

One more problem with 3. is that multiple begin stream could run in parallel for the same module, and all could exhaust the memory halfway through, resulting in no streams using the GPU.

**Comment by fwyzard** [comment]
To summarise, I don't have a good suggestion at the moment...

**Comment by makortel** [comment]
The functionality of option 2 could also be implemented by as a configuration parameter to `CUDAService`, which then `GPUCuda` would check here
https://github.com/cms-patatrack/cmssw/blob/b2055cd9628a5f0d64687bc11faea8fe46b0af99/HeterogeneousCore/CUDACore/src/GPUCuda.cc#L27-L30
Then at least there would be a single configuration point, and all modules would honor the limit automatically.

**Comment by makortel** [comment]
From configuration side, would the following be good enough (aiming to be consistent with the number of treads and streams)?
```python
procesws.CUDAService.maxNumberOfEdmStreams = cms.untracked.uint32(0)
# 0 means no limit
# >= 1 means a limit
```
Should this actually be a per-device setting? (let's ignore the case where a machine has very differently performing GPUs) Then I' call it something along `maxNumberOfEdmStreamsPerDevice`. I suspect though that either way the user has to do some math between this parameter, number of devices, and the number of edm::Streams.

**Comment by fwyzard** [comment]
`maxNumberOfEdmStreamsPerDevice` sounds good

**Comment by fwyzard** [comment]
actually, `numberOfStreamsPerDevice` would be a bit simpler

**Comment by makortel** [comment]
> actually, `numberOfStreamsPerDevice` would be a bit simpler

Yes, but would it be clear that the parameter refers to edm streams and not CUDA streams?

**Comment by fwyzard** [comment]
I'm tempted to answer "yes" :-)

`numberOfEdmStreamsPerDevice` could work as well, of course

**Comment by makortel** [comment]
Ok, let's go for the shorter one then (`numberOfStreamsPerDevice`).
